### PR TITLE
perf(zk): merge parallelized parts related to grand product

### PR DESCRIPTION
**Pull request by dongchangYoo**
In this PR, the sections where `base::Parallelize()` was called multiple times have been modified so that each is called only once. Unfortunately, this PR does not result in any current performance changes, but it has improved code readability and made it easier to optimize parallelization in the future.

### CODE REVIEW

File: tachyon/zk/plonk/permutation/permutation_prover.h [hunk 1]
```diff
@@ -116,14 +116,12 @@ class PermutationProver {
...
-  static std::function<base::ParallelizeCallback3<F>(size_t)>
-  CreateNumeratorCallback(
+  static std::function<F(size_t, size_t)> CreateNumeratorCallback(
       const std::vector<base::Ref<const Evals>>& unpermuted_columns,
       const std::vector<base::Ref<const Evals>>& value_columns, const F& beta,
       const F& gamma);
 
-  static std::function<base::ParallelizeCallback3<F>(size_t)>
-  CreateDenominatorCallback(
+  static std::function<F(size_t, size_t)> CreateDenominatorCallback(
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/permutation_prover.h:124]: Could you remove `tachyon/base/parallelize.h` in this file?

File: tachyon/zk/plonk/permutation/permutation_prover.h [hunk 2]
```diff
@@ -116,14 +116,12 @@ class PermutationProver {
   void Evaluate(ProverBase<PCS>* prover,
                 const PermutationOpeningPointSet<F>& point_set) const;
 
-  static std::function<base::ParallelizeCallback3<F>(size_t)>
-  CreateNumeratorCallback(
+  static std::function<F(size_t, size_t)> CreateNumeratorCallback(
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/permutation_prover.h:119]:
```suggestion
  static std::function<F(size_t, RowIndex)> CreateNumeratorCallback(
```

File: tachyon/zk/plonk/permutation/grand_product_argument.h [hunk 3]
```diff
@@ -52,14 +56,26 @@ class GrandProductArgument {
...
-    for (size_t i = 0; i < num_cols; ++i) {
-      base::Parallelize(grand_product, numerator_callback(i));
+    size_t chunk_size = base::GetNumElementsPerThread(grand_product);
+    size_t num_chunks = (size + chunk_size - 1) / chunk_size;
+
+    OPENMP_PARALLEL_FOR(size_t i = 0; i < num_chunks; ++i) {
+      size_t start = i * chunk_size;
+      size_t end = i == num_chunks - 1 ? size : start + chunk_size;
+      for (size_t j = 0; j < num_cols; ++j) {
+        for (size_t k = start; k < end; ++k) {
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/grand_product_argument.h:66]:
```suggestion
        for (RowIndex k = start; k < end; ++k) {
```

File: tachyon/zk/plonk/permutation/grand_product_argument.h [hunk 4]
```diff
@@ -52,14 +56,26 @@ class GrandProductArgument {
...
-    CHECK(F::BatchInverseInPlace(grand_product));
-
-    for (size_t i = 0; i < num_cols; ++i) {
-      base::Parallelize(grand_product, numerator_callback(i));
+    size_t chunk_size = base::GetNumElementsPerThread(grand_product);
+    size_t num_chunks = (size + chunk_size - 1) / chunk_size;
+
+    OPENMP_PARALLEL_FOR(size_t i = 0; i < num_chunks; ++i) {
+      size_t start = i * chunk_size;
+      size_t end = i == num_chunks - 1 ? size : start + chunk_size;
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/grand_product_argument.h:64]:
```suggestion
      RowIndex start = i * chunk_size;
      RowIndex end = i == num_chunks - 1 ? size : start + chunk_size;
```

File: tachyon/zk/plonk/permutation/grand_product_argument.h [hunk 5]
```diff
@@ -52,14 +56,26 @@ class GrandProductArgument {
...
+        for (size_t k = start; k < end; ++k) {
+          grand_product[k] *= denominator_callback(j, k);
+        }
+      }
+
+      auto grand_subspan = grand_product.subspan(start, end - start);
+      CHECK(F::BatchInverseInPlaceSerial(grand_subspan));
+
+      for (size_t j = 0; j < num_cols; ++j) {
+        for (size_t k = start; k < end; ++k) {
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/grand_product_argument.h:75]:
```suggestion
        for (RowIndex k = start; k < end; ++k) {
``

File: tachyon/zk/plonk/permutation/grand_product_argument.h [hunk 6]
```diff
@@ -26,11 +26,15 @@ class GrandProductArgument {
...
-    base::Parallelize(grand_product, std::move(denominator_callback));
+    for (size_t i = 0; i < size; ++i) {
+      grand_product[i] = denominator_callback(i);
+    }
 
-    CHECK(F::BatchInverseInPlace(grand_product));
+    CHECK(F::BatchInverseInPlaceSerial(grand_product));
 
-    base::Parallelize(grand_product, std::move(numerator_callback));
+    for (size_t i = 0; i < size; ++i) {
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/grand_product_argument.h:35]:
```suggestion
    for (RowIndex i = 0; i < size; ++i) {
``

File: tachyon/zk/plonk/permutation/grand_product_argument.h [hunk 7]
```diff
@@ -26,11 +26,15 @@ class GrandProductArgument {
     std::vector<F> z = base::CreateVector(size + 1, F::Zero());
     absl::Span<F> grand_product = absl::MakeSpan(z).subspan(1);
 
-    base::Parallelize(grand_product, std::move(denominator_callback));
+    for (size_t i = 0; i < size; ++i) {
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/grand_product_argument.h:29]:
```suggestion
    for (RowIndex i = 0; i < size; ++i) {
``

File: tachyon/zk/lookup/halo2/prover_impl.h [hunk 8]
```diff
@@ -242,36 +240,26 @@ void Prover<Poly, Evals>::Open(
...
+  return [&compressed_pair, &beta, &gamma](RowIndex row_index) {
+    return (compressed_pair.input()[row_index] + beta) *
+           (compressed_pair.table()[row_index] + gamma);
   };
 }
 
 // static
 template <typename Poly, typename Evals>
-base::ParallelizeCallback3<typename Poly::Field>
+std::function<typename Poly::Field(size_t)>
```
Comment (chokobole) [inline tachyon/zk/lookup/halo2/prover_impl.h:255]:
```suggestion
std::function<typename Poly::Field(RowIndex)>
```

File: tachyon/zk/lookup/halo2/prover_impl.h [hunk 9]
```diff
@@ -242,36 +240,26 @@ void Prover<Poly, Evals>::Open(
...
```
Comment (chokobole) [inline tachyon/zk/lookup/halo2/prover_impl.h:243]:
```suggestion
std::function<typename Poly::Field(RowIndex)>
```

File: tachyon/zk/lookup/halo2/prover_impl.h [hunk 10]
```diff
@@ -242,36 +240,26 @@ void Prover<Poly, Evals>::Open(
...
-             absl::Span<F> chunk, size_t chunk_index, size_t chunk_size) {
-    size_t i = chunk_index * chunk_size;
-    for (F& value : chunk) {
-      value = (permuted_pair.input().evals()[i] + beta) *
-              (permuted_pair.table().evals()[i] + gamma);
-      ++i;
-    }
+    const F& beta,
+    const F& gamma) {  // (A'(xᵢ) + β) * (S'(xᵢ) + γ)
+  return [&permuted_pair, &beta, &gamma](RowIndex row_index) {
```
Comment (chokobole) [inline tachyon/zk/lookup/halo2/prover_impl.h:260]:
```suggestion
    const F& gamma) {  
  // (A'(xᵢ) + β) * (S'(xᵢ) + γ)    
  return [&permuted_pair, &beta, &gamma](RowIndex row_index) {
```

File: tachyon/zk/lookup/halo2/prover.h [hunk 11]
```diff
@@ -152,10 +152,10 @@ class Prover {
   void Evaluate(ProverBase<PCS>* prover,
                 const OpeningPointSet<F>& point_set) const;
 
-  static base::ParallelizeCallback3<F> CreateNumeratorCallback(
+  static std::function<F(size_t)> CreateNumeratorCallback(
       const LookupPair<Evals>& compressed_pair, const F& beta, const F& gamma);
 
-  static base::ParallelizeCallback3<F> CreateDenominatorCallback(
+  static std::function<F(size_t)> CreateDenominatorCallback(
```
Comment (chokobole) [inline tachyon/zk/lookup/halo2/prover.h:158]:
```suggestion
  static std::function<F(RowIndex)> CreateDenominatorCallback(
```

File: tachyon/zk/lookup/halo2/prover.h [hunk 12]
```diff
@@ -152,10 +152,10 @@ class Prover {
   void Evaluate(ProverBase<PCS>* prover,
                 const OpeningPointSet<F>& point_set) const;
 
-  static base::ParallelizeCallback3<F> CreateNumeratorCallback(
+  static std::function<F(size_t)> CreateNumeratorCallback(
```
Comment (chokobole) [inline tachyon/zk/lookup/halo2/prover.h:155]:
```suggestion
  static std::function<F(RowIndex)> CreateNumeratorCallback(
```

File: tachyon/zk/plonk/permutation/permutation_prover_impl.h [hunk 13]
```diff
@@ -192,47 +193,33 @@ void PermutationProver<Poly, Evals>::OpenPermutationProvingKey(
 
 // static
 template <typename Poly, typename Evals>
-std::function<base::ParallelizeCallback3<typename Poly::Field>(size_t)>
+std::function<typename Poly::Field(size_t, size_t)>
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/permutation_prover_impl.h:196]:
```suggestion
std::function<typename Poly::Field(size_t, RowIndex)>
```

File: tachyon/zk/plonk/permutation/permutation_prover_impl.h [hunk 14]
```diff
@@ -192,47 +193,33 @@ void PermutationProver<Poly, Evals>::OpenPermutationProvingKey(
...
-      }
-    };
+    return values[row_index] + beta * unpermuted_values[row_index] + gamma;
   };
 }
 
 // static
 template <typename Poly, typename Evals>
-std::function<base::ParallelizeCallback3<typename Poly::Field>(size_t)>
+std::function<typename Poly::Field(size_t, size_t)>
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/permutation_prover_impl.h:212]:
```suggestion
std::function<typename Poly::Field(size_t, RowIndex)>
```

File: tachyon/zk/plonk/permutation/permutation_prover.h [hunk 15]
```diff
@@ -12,8 +12,8 @@
 #include <functional>
 #include <vector>
 
+#include "tachyon/base/functional/functor_traits.h"
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/permutation_prover.h:15]: Please remove this

**Review by chokobole — APPROVED** [review]
LGTM

**Review by ashjeong — APPROVED** [review]
LGTM

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Pull request merged**